narwhal/consensus/src/bullshark.rs

```diff
@@ -314,4 +329,46 @@ impl Bullshark {
         // Return its certificate and the certificate's digest.
         dag.get(&round).and_then(|x| x.get(&leader))
     }
+
+    /// Updates and calculates the reputation score for the current commit managing any internal state.
+    /// It returns the updated reputation score.
+    fn update_reputation_score(
+        &mut self,
+        state: &mut ConsensusState,
+        committed_sequence: &[Certificate],
+        sub_dag_index: u64,
+    ) -> ReputationScores {
+        // we reset the scores for every schedule change window.
+        // TODO: when schedule change is implemented we should probably change a little bit
+        // this logic here.
+        if sub_dag_index % self.change_schedule_every_committed_sub_dags == 0 {
+            state.last_consensus_reputation_score = ReputationScores::default()
+        }
+
+        // update the score for the previous leader. If no previous leader exists,
+        // then this is the first time we commit a leader, so no score update takes place
+        if let Some(previous_leader) = state.last_committed_leader {
+            for certificate in committed_sequence {
+                // TODO: we could iterate only the certificates of the round above the previous leader's round
+                if certificate
+                    .header
+                    .parents
+                    .iter()
+                    .any(|digest| *digest == previous_leader)
+                {
+                    state
+                        .last_consensus_reputation_score
+                        .add_score(certificate.origin(), 1);
+                }
+            }
+        }
+
+        // we check if this is the last subdag of the current schedule. If yes then we mark the
+        // scores as final_of_schedule = true so any downstream user can now that those are the last
+        // ones calculated for the current schedule.
```

Comment:
How will final_of_schedule be used?

Reply:
The reputation scores get updated per commit and are published as part of the `CommitedSubdag` , which is consumed from the `consensus_handler` in SUI. Now, if on SUI we want to use the scores, we can't/shouldn't use them in a streaming fashion - meaning update them locally every time we receive a new subdag - but rather use the "final" score that has been calculated from the current schedule/window. That will allow us to take decisions more deterministically as we avoid jittery behaviour where validators are on different commits and see different scores/results. Now, in order to to help SUI to know _when_ to update/apply the new scores, I've introduced the `final_of_schedule` bool as an easy/agnostic way for them to go and update the schedule only when this is `true`.